File: scripts/bib_to_yaml.py
import re
import unicodedata
from pathlib import Path
from datetime import date

import bibtexparser
from bibtexparser.bparser import BibTexParser
from bibtexparser.customization import homogenize_latex_encoding
import yaml

# === Paths ===
ROOT = Path(__file__).resolve().parents[1]
BIB_DIR = ROOT / "bib"
DATA_DIR = ROOT / "data"

# Define which bib files to convert
BIB_FILES = {
    "first_pub.bib": "first_pub.yaml",
    "co_pub.bib": "co_pub.yaml",
}

# === Constants ===
MONTH_MAP = {
    "jan": 1, "feb": 2, "mar": 3, "apr": 4,
    "may": 5, "jun": 6, "jul": 7, "aug": 8,
    "sep": 9, "oct": 10, "nov": 11, "dec": 12,
}

JOURNAL_MAP = {
    r"\apj": "ApJ",
    r"\mnras": "MNRAS",
    r"\aap": "A&A",
    r"\aj": "AJ",
    r"\apjs": "ApJS",
    r"\apjl": "ApJL",
}


# === Utility functions ===
def load_bib(path: Path):
    parser = BibTexParser(common_strings=True)
    parser.customization = homogenize_latex_encoding
    with open(path, encoding="utf-8") as f:
        return bibtexparser.load(f, parser=parser)


def parse_month(m):
    if not m:
        return 1
    m = str(m).strip().lower()
    try:
        mi = int(m)
        if 1 <= mi <= 12:
            return mi
    except ValueError:
        pass
    return MONTH_MAP.get(m[:3], 1)


def normalize_journal(journal):
    if not journal:
        return ""
    j = str(journal).strip()
    if j.startswith("{") and j.endswith("}"):
        j = j[1:-1].strip()
    return JOURNAL_MAP.get(j, j)


def clean_latex(s: str) -> str:
    """Remove BibTeX/LaTeX braces and escaped accents."""
    if not s:
        return ""
    s = re.sub(r"[{}]", "", s)
    latex_accents = {
        r"\\'a": "á", r"\\'e": "é", r"\\'i": "í", r"\\'o": "ó", r"\\'u": "ú",
        r"\\`a": "à", r"\\`e": "è", r"\\`i": "ì", r"\\`o": "ò", r"\\`u": "ù",
        r'\\"a': "ä", r'\\"o': "ö", r'\\"u': "ü", r'\\"A': "Ä", r'\\"O': "Ö", r'\\"U': "Ü",
        r"\\~n": "ñ", r"\\'A": "Á", r"\\'E": "É", r"\\'I": "Í", r"\\'O": "Ó", r"\\'U": "Ú",
        r"\\c{c}": "ç", r"\\c{C}": "Ç",
        r"\\ss": "ß",
    }
    for k, v in latex_accents.items():
        s = re.sub(k, v, s)
    s = re.sub(r"\\[a-zA-Z]+", "", s)
    s = unicodedata.normalize("NFC", s)
    s = re.sub(r"\s+", " ", s).strip()
    return s


# === Core conversion ===
def entry_to_yaml_item(entry: dict) -> dict:
    title = clean_latex(entry.get("title", "").strip())
    authors_raw = entry.get("author", "")
    authors = clean_latex(authors_raw.replace(" and ", "; "))

    journal = entry.get("journal") or entry.get("booktitle") or ""
    venue = clean_latex(normalize_journal(journal))

    try:
        year = int(entry.get("year", "1900"))
    except ValueError:
        year = 1900

    month = parse_month(entry.get("month"))
    date_str = f"{year:04d}-{month:02d}-01"

    # arXiv
    arxiv_url = ""
    if entry.get("archivePrefix", "").lower() == "arxiv":
        eprint = entry.get("eprint", "").strip()
        if eprint:
            arxiv_url = f"https://arxiv.org/abs/{eprint}"

    # DOI
    doi_url = ""
    doi = (entry.get("doi") or "").strip()
    if doi:
        doi_url = doi if doi.lower().startswith("http") else f"https://doi.org/{doi}"

    # ADS
    ads_url = (entry.get("adsurl") or "").strip()
    if ads_url and not ads_url.startswith("http"):
        ads_url = f"https://{ads_url}"

    item = {
        "title": title,
        "authors": authors,
        "venue": venue,
        "year": year,
        "date": date_str,
        "links": {
            "arxiv": arxiv_url,
            "doi": doi_url,
            "ads": ads_url,
            "pdf": "",
        },
    }
    return item


def convert_bib_to_yaml(bib_path: Path, yaml_path: Path):
    if not bib_path.exists():
        print(f"⚠️  Skipping missing file: {bib_path}")
        return

    bibdb = load_bib(bib_path)
    items = [entry_to_yaml_item(e) for e in bibdb.entries]
    items.sort(key=lambda x: x["date"], reverse=True)

    with open(yaml_path, "w", encoding="utf-8") as f:
        yaml.safe_dump(items, f, sort_keys=False, allow_unicode=True)

    print(f"✅ Converted {bib_path.name} → {yaml_path.name} ({len(items)} entries)")


def main():
    for bib_name, yaml_name in BIB_FILES.items():
        bib_path = BIB_DIR / bib_name
        yaml_path = DATA_DIR / yaml_name
        convert_bib_to_yaml(bib_path, yaml_path)


if __name__ == "__main__":
    main()

File: scripts/bib_to_yaml.mjs
/**
 * BibTeX-to-YAML converter for publications.
 *
 * Reads .bib files from bib/ directory and writes a merged, sorted
 * publications.yaml to src/data/publications.yaml.
 *
 * Usage: node scripts/bib_to_yaml.mjs
 */

import { parse } from "@retorquere/bibtex-parser";
import yaml from "js-yaml";
import fs from "node:fs";
import path from "node:path";

const ROOT = path.resolve(import.meta.dirname, "..");
const BIB_DIR = path.join(ROOT, "bib");
const OUTPUT = path.join(ROOT, "src", "data", "publications.yaml");

// === Constants ===

const MONTH_MAP = {
  jan: 1, feb: 2, mar: 3, apr: 4,
  may: 5, jun: 6, jul: 7, aug: 8,
  sep: 9, oct: 10, nov: 11, dec: 12,
};

const JOURNAL_MAP = {
  "\\apj": "ApJ",
  "\\mnras": "MNRAS",
  "\\aap": "A&A",
  "\\aj": "AJ",
  "\\apjs": "ApJS",
  "\\apjl": "ApJL",
};

// Words to skip when generating the slug from a title
const STOP_WORDS = new Set([
  "a", "an", "the", "of", "on", "in", "for", "and", "or", "to",
  "from", "with", "by", "at", "its", "via", "using",
]);

// === Utility functions ===

/**
 * Remove BibTeX/LaTeX braces and convert LaTeX accents to Unicode.
 */
function cleanLatex(s) {
  if (!s) return "";

  // Remove braces
  s = s.replace(/[{}]/g, "");

  // LaTeX accent map
  const accents = [
    [/\\'a/g, "\u00e1"], [/\\'e/g, "\u00e9"], [/\\'i/g, "\u00ed"],
    [/\\'o/g, "\u00f3"], [/\\'u/g, "\u00fa"], [/\\'A/g, "\u00c1"],
    [/\\'E/g, "\u00c9"], [/\\'I/g, "\u00cd"], [/\\'O/g, "\u00d3"],
    [/\\'U/g, "\u00da"], [/\\`a/g, "\u00e0"], [/\\`e/g, "\u00e8"],
    [/\\`i/g, "\u00ec"], [/\\`o/g, "\u00f2"], [/\\`u/g, "\u00f9"],
    [/\\"a/g, "\u00e4"], [/\\"o/g, "\u00f6"], [/\\"u/g, "\u00fc"],
    [/\\"A/g, "\u00c4"], [/\\"O/g, "\u00d6"], [/\\"U/g, "\u00dc"],
    [/\\~n/g, "\u00f1"], [/\\cc/g, "\u00e7"], [/\\cC/g, "\u00c7"],
    [/\\ss/g, "ss"],
  ];

  for (const [pattern, replacement] of accents) {
    s = s.replace(pattern, replacement);
  }

  // Remove remaining LaTeX commands
  s = s.replace(/\\[a-zA-Z]+/g, "");

  // Normalize Unicode
  s = s.normalize("NFC");

  // Collapse whitespace
  s = s.replace(/\s+/g, " ").trim();

  return s;
}

/**
 * Convert month string or number to integer 1-12.
 */
function parseMonth(m) {
  if (!m) return 1;
  const s = String(m).trim().toLowerCase();
  const n = parseInt(s, 10);
  if (!isNaN(n) && n >= 1 && n <= 12) return n;
  return MONTH_MAP[s.slice(0, 3)] || 1;
}

/**
 * Map LaTeX journal macros to standard abbreviations.
 */
function normalizeJournal(journal) {
  if (!journal) return "";
  let j = String(journal).trim();
  if (j.startsWith("{") && j.endsWith("}")) {
    j = j.slice(1, -1).trim();
  }
  return JOURNAL_MAP[j] || j;
}

/**
 * Generate a slug ID from first author surname, year, and first significant title word.
 */
function generateId(authorLastName, year, title) {
  const surname = authorLastName.toLowerCase().replace(/[^a-z]/g, "");
  const words = title.toLowerCase().replace(/[^a-z0-9\s]/g, "").split(/\s+/);
  const significantWord = words.find((w) => w.length > 0 && !STOP_WORDS.has(w)) || "untitled";
  return `${surname}-${year}-${significantWord}`;
}

/**
 * Format authors from parsed BibTeX author objects to "Last, First; Last, First" string.
 */
function formatAuthors(authors) {
  if (!authors || !Array.isArray(authors)) return "";
  return authors
    .map((a) => {
      const last = cleanLatex(a.lastName || "");
      const first = cleanLatex(a.firstName || "");
      return first ? `${last}, ${first}` : last;
    })
    .join("; ");
}

/**
 * Convert a parsed BibTeX entry to the YAML publication object format.
 */
function entryToYaml(entry, role) {
  const fields = entry.fields;

  const title = cleanLatex(fields.title || "");
  const authors = formatAuthors(fields.author);
  const venue = cleanLatex(normalizeJournal(fields.journal || fields.booktitle || ""));
  const year = parseInt(fields.year, 10) || 1900;
  const month = parseMonth(fields.month);
  const date = `${String(year).padStart(4, "0")}-${String(month).padStart(2, "0")}-01`;

  // Extract first author last name for ID
  const firstAuthorLast = fields.author?.[0]?.lastName || "unknown";

  const id = generateId(firstAuthorLast, year, fields.title || "");

  // arXiv URL
  let arxiv = "";
  if (String(fields.archiveprefix || "").toLowerCase() === "arxiv" && fields.eprint) {
    arxiv = `https://arxiv.org/abs/${fields.eprint.trim()}`;
  }

  // DOI URL
  let doi = "";
  const rawDoi = (fields.doi || "").trim();
  if (rawDoi) {
    doi = rawDoi.toLowerCase().startsWith("http") ? rawDoi : `https://doi.org/${rawDoi}`;
  }

  // ADS URL
  let ads = (fields.adsurl || "").trim();
  if (ads && !ads.startsWith("http")) {
    ads = `https://${ads}`;
  }

  return {
    id,
    title,
    authors,
    venue,
    year,
    date,
    role,
    links: { arxiv, doi, ads, pdf: "" },
  };
}

/**
 * Parse a .bib file and return publication entries.
 */
function parseBibFile(filePath, role) {
  if (!fs.existsSync(filePath)) {
    console.warn(`Warning: ${filePath} not found, skipping.`);
    return [];
  }

  const content = fs.readFileSync(filePath, "utf-8");
  const result = parse(content, { sentenceCase: false });

  return result.entries.map((entry) => entryToYaml(entry, role));
}

/**
 * Main: read bib files, merge, sort by date descending, write YAML.
 */
function convertBibToYaml() {
  const firstPubPath = path.join(BIB_DIR, "first_pub.bib");
  const coPubPath = path.join(BIB_DIR, "co_pub.bib");

  const firstPubs = parseBibFile(firstPubPath, "first-author");
  const coPubs = parseBibFile(coPubPath, "co-author");

  const all = [...firstPubs, ...coPubs];
  all.sort((a, b) => b.date.localeCompare(a.date));

  // Ensure output directory exists
  const outDir = path.dirname(OUTPUT);
  if (!fs.existsSync(outDir)) {
    fs.mkdirSync(outDir, { recursive: true });
  }

  const yamlStr = yaml.dump(all, { sortKeys: false, lineWidth: -1 });
  fs.writeFileSync(OUTPUT, yamlStr, "utf-8");

  console.log(`Converted ${all.length} publications to ${OUTPUT}`);
  console.log(`  First-author: ${firstPubs.length}`);
  console.log(`  Co-author: ${coPubs.length}`);
}

convertBibToYaml();
